File: stripe/src/test/java/com/stripe/android/AbsFakeStripeRepository.java
package com.stripe.android;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.stripe.android.exception.APIException;
import com.stripe.android.model.ConfirmPaymentIntentParams;
import com.stripe.android.model.ConfirmSetupIntentParams;
import com.stripe.android.model.Customer;
import com.stripe.android.model.PaymentIntent;
import com.stripe.android.model.PaymentMethod;
import com.stripe.android.model.PaymentMethodCreateParams;
import com.stripe.android.model.SetupIntent;
import com.stripe.android.model.ShippingInformation;
import com.stripe.android.model.Source;
import com.stripe.android.model.SourceParams;
import com.stripe.android.model.Stripe3ds2AuthResult;
import com.stripe.android.model.Token;

import java.util.Collections;
import java.util.List;
import java.util.Map;

abstract class AbsFakeStripeRepository implements StripeRepository {

    @Nullable
    @Override
    public PaymentIntent confirmPaymentIntent(
            @NonNull ConfirmPaymentIntentParams confirmPaymentIntentParams,
            @NonNull ApiRequest.Options options) {
        return null;
    }

    @Nullable
    @Override
    public PaymentIntent retrievePaymentIntent(
            @NonNull String clientSecret,
            @NonNull ApiRequest.Options options) {
        return null;
    }

    @Nullable
    @Override
    public SetupIntent confirmSetupIntent(
            @NonNull ConfirmSetupIntentParams confirmSetupIntentParams,
            @NonNull ApiRequest.Options options) {
        return null;
    }

    @Nullable
    @Override
    public SetupIntent retrieveSetupIntent(
            @NonNull String clientSecret, @NonNull ApiRequest.Options options) {
        return null;
    }

    @Nullable
    @Override
    public Source createSource(
            @NonNull SourceParams sourceParams, @NonNull ApiRequest.Options options) {
        return null;
    }

    @Nullable
    @Override
    public Source retrieveSource(
            @NonNull String sourceId, @NonNull String clientSecret,
            @NonNull ApiRequest.Options options) {
        return null;
    }

    @Nullable
    @Override
    public PaymentMethod createPaymentMethod(
            @NonNull PaymentMethodCreateParams paymentMethodCreateParams,
            @NonNull ApiRequest.Options options) {
        return null;
    }

    @Nullable
    @Override
    public Token createToken(
            @NonNull Map<String, ?> tokenParams, @NonNull ApiRequest.Options options,
            @NonNull String tokenType) {
        return null;
    }

    @Nullable
    @Override
    public Source addCustomerSource(
            @NonNull String customerId, @NonNull String publishableKey,
            @NonNull List<String> productUsageTokens, @NonNull String sourceId,
            @NonNull String sourceType, @NonNull ApiRequest.Options requestOptions) throws APIException {
        return null;
    }

    @Nullable
    @Override
    public Source deleteCustomerSource(
            @NonNull String customerId, @NonNull String publishableKey,
            @NonNull List<String> productUsageTokens, @NonNull String sourceId,
            @NonNull ApiRequest.Options requestOptions) throws APIException {
        return null;
    }

    @Nullable
    @Override
    public PaymentMethod attachPaymentMethod(
            @NonNull String customerId, @NonNull String publishableKey,
            @NonNull List<String> productUsageTokens, @NonNull String paymentMethodId,
            @NonNull ApiRequest.Options requestOptions) throws APIException {
        return null;
    }

    @Nullable
    @Override
    public PaymentMethod detachPaymentMethod(
            @NonNull String publishableKey, @NonNull List<String> productUsageTokens,
            @NonNull String paymentMethodId, @NonNull ApiRequest.Options requestOptions) throws APIException {
        return null;
    }

    @NonNull
    @Override
    public List<PaymentMethod> getPaymentMethods(
            @NonNull String customerId, @NonNull String paymentMethodType,
            @NonNull String publishableKey, @NonNull List<String> productUsageTokens,
            @NonNull ApiRequest.Options requestOptions) throws APIException {
        return Collections.emptyList();
    }

    @Nullable
    @Override
    public Customer setDefaultCustomerSource(
            @NonNull String customerId, @NonNull String publishableKey,
            @NonNull List<String> productUsageTokens, @NonNull String sourceId,
            @NonNull String sourceType, @NonNull ApiRequest.Options requestOptions) throws APIException {
        return null;
    }

    @Nullable
    @Override
    public Customer setCustomerShippingInfo(
            @NonNull String customerId,
            @NonNull String publishableKey,
            @NonNull List<String> productUsageTokens,
            @NonNull ShippingInformation shippingInformation,
            @NonNull ApiRequest.Options requestOptions) {
        return null;
    }

    @Nullable
    @Override
    public Customer retrieveCustomer(
            @NonNull String customerId, @NonNull ApiRequest.Options requestOptions) {
        return null;
    }

    @NonNull
    @Override
    public String retrieveIssuingCardPin(
            @NonNull String cardId, @NonNull String verificationId,
            @NonNull String userOneTimeCode, @NonNull String ephemeralKeySecret) {
        return "";
    }

    @Override
    public void updateIssuingCardPin(
            @NonNull String cardId, @NonNull String newPin, @NonNull String verificationId,
            @NonNull String userOneTimeCode, @NonNull String ephemeralKeySecret) {
    }

    @Override
    public void start3ds2Auth(
            @NonNull Stripe3ds2AuthParams authParams, @NonNull String stripeIntentId,
            @NonNull ApiRequest.Options requestOptions,
            @NonNull ApiResultCallback<Stripe3ds2AuthResult> callback) {
    }

    @Override
    public void complete3ds2Auth(
            @NonNull String sourceId, @NonNull ApiRequest.Options requestOptions,
            @NonNull ApiResultCallback<Boolean> callback) {
    }
}

File: stripe/src/test/java/com/stripe/android/AbsFakeStripeRepository.kt
package com.stripe.android

import com.stripe.android.cards.Bin
import com.stripe.android.exception.APIException
import com.stripe.android.model.BinFixtures
import com.stripe.android.model.CardMetadata
import com.stripe.android.model.Complete3ds2Result
import com.stripe.android.model.ConfirmPaymentIntentParams
import com.stripe.android.model.ConfirmSetupIntentParams
import com.stripe.android.model.Customer
import com.stripe.android.model.FpxBankStatuses
import com.stripe.android.model.ListPaymentMethodsParams
import com.stripe.android.model.PaymentIntent
import com.stripe.android.model.PaymentMethod
import com.stripe.android.model.PaymentMethodCreateParams
import com.stripe.android.model.SetupIntent
import com.stripe.android.model.ShippingInformation
import com.stripe.android.model.Source
import com.stripe.android.model.SourceParams
import com.stripe.android.model.Stripe3ds2AuthParams
import com.stripe.android.model.Stripe3ds2AuthResultFixtures
import com.stripe.android.model.StripeFile
import com.stripe.android.model.StripeFileParams
import com.stripe.android.model.StripeIntent
import com.stripe.android.model.Token
import com.stripe.android.model.TokenParams
import org.json.JSONObject

internal abstract class AbsFakeStripeRepository : StripeRepository {

    override fun confirmPaymentIntent(
        confirmPaymentIntentParams: ConfirmPaymentIntentParams,
        options: ApiRequest.Options,
        expandFields: List<String>
    ): PaymentIntent? {
        return null
    }

    override fun retrievePaymentIntent(
        clientSecret: String,
        options: ApiRequest.Options,
        expandFields: List<String>
    ): PaymentIntent? {
        return null
    }

    override fun cancelPaymentIntentSource(
        paymentIntentId: String,
        sourceId: String,
        options: ApiRequest.Options
    ): PaymentIntent? {
        return null
    }

    override fun confirmSetupIntent(
        confirmSetupIntentParams: ConfirmSetupIntentParams,
        options: ApiRequest.Options,
        expandFields: List<String>
    ): SetupIntent? {
        return null
    }

    override fun retrieveSetupIntent(
        clientSecret: String,
        options: ApiRequest.Options,
        expandFields: List<String>
    ): SetupIntent? {
        return null
    }

    override fun cancelSetupIntentSource(
        setupIntentId: String,
        sourceId: String,
        options: ApiRequest.Options
    ): SetupIntent? {
        return null
    }

    override fun retrieveIntent(
        clientSecret: String,
        options: ApiRequest.Options,
        expandFields: List<String>,
        callback: ApiResultCallback<StripeIntent>
    ) {
    }

    override fun cancelIntent(
        stripeIntent: StripeIntent,
        sourceId: String,
        options: ApiRequest.Options,
        callback: ApiResultCallback<StripeIntent>
    ) {
    }

    override fun createSource(
        sourceParams: SourceParams,
        options: ApiRequest.Options
    ): Source? {
        return null
    }

    override fun retrieveSource(
        sourceId: String,
        clientSecret: String,
        options: ApiRequest.Options
    ): Source? {
        return null
    }

    override fun retrieveSource(
        sourceId: String,
        clientSecret: String,
        options: ApiRequest.Options,
        callback: ApiResultCallback<Source>
    ) {
    }

    override fun createPaymentMethod(
        paymentMethodCreateParams: PaymentMethodCreateParams,
        options: ApiRequest.Options
    ): PaymentMethod? {
        return null
    }

    override fun createToken(
        tokenParams: TokenParams,
        options: ApiRequest.Options
    ): Token? {
        return null
    }

    @Throws(APIException::class)
    override fun addCustomerSource(
        customerId: String,
        publishableKey: String,
        productUsageTokens: Set<String>,
        sourceId: String,
        sourceType: String,
        requestOptions: ApiRequest.Options
    ): Source? {
        return null
    }

    @Throws(APIException::class)
    override fun deleteCustomerSource(
        customerId: String,
        publishableKey: String,
        productUsageTokens: Set<String>,
        sourceId: String,
        requestOptions: ApiRequest.Options
    ): Source? {
        return null
    }

    @Throws(APIException::class)
    override fun attachPaymentMethod(
        customerId: String,
        publishableKey: String,
        productUsageTokens: Set<String>,
        paymentMethodId: String,
        requestOptions: ApiRequest.Options
    ): PaymentMethod? {
        return null
    }

    @Throws(APIException::class)
    override fun detachPaymentMethod(
        publishableKey: String,
        productUsageTokens: Set<String>,
        paymentMethodId: String,
        requestOptions: ApiRequest.Options
    ): PaymentMethod? {
        return null
    }

    @Throws(APIException::class)
    override fun getPaymentMethods(
        listPaymentMethodsParams: ListPaymentMethodsParams,
        publishableKey: String,
        productUsageTokens: Set<String>,
        requestOptions: ApiRequest.Options
    ): List<PaymentMethod> {
        return emptyList()
    }

    @Throws(APIException::class)
    override fun setDefaultCustomerSource(
        customerId: String,
        publishableKey: String,
        productUsageTokens: Set<String>,
        sourceId: String,
        sourceType: String,
        requestOptions: ApiRequest.Options
    ): Customer? {
        return null
    }

    override fun setCustomerShippingInfo(
        customerId: String,
        publishableKey: String,
        productUsageTokens: Set<String>,
        shippingInformation: ShippingInformation,
        requestOptions: ApiRequest.Options
    ): Customer? {
        return null
    }

    override fun retrieveCustomer(
        customerId: String,
        productUsageTokens: Set<String>,
        requestOptions: ApiRequest.Options
    ): Customer? {
        return null
    }

    override fun retrieveIssuingCardPin(
        cardId: String,
        verificationId: String,
        userOneTimeCode: String,
        ephemeralKeySecret: String
    ): String {
        return ""
    }

    override fun updateIssuingCardPin(
        cardId: String,
        newPin: String,
        verificationId: String,
        userOneTimeCode: String,
        ephemeralKeySecret: String
    ) {
    }

    override suspend fun getFpxBankStatus(
        options: ApiRequest.Options
    ) = FpxBankStatuses()

    override suspend fun getCardMetadata(bin: Bin, options: ApiRequest.Options) =
        CardMetadata(
            BinFixtures.VISA,
            emptyList()
        )

    override suspend fun start3ds2Auth(
        authParams: Stripe3ds2AuthParams,
        stripeIntentId: String,
        requestOptions: ApiRequest.Options
    ) = Stripe3ds2AuthResultFixtures.ARES_CHALLENGE_FLOW

    override suspend fun complete3ds2Auth(
        sourceId: String,
        requestOptions: ApiRequest.Options
    ): Complete3ds2Result = Complete3ds2Result(true)

    override fun createFile(
        fileParams: StripeFileParams,
        requestOptions: ApiRequest.Options
    ): StripeFile {
        return StripeFile()
    }

    override fun retrieveObject(url: String, requestOptions: ApiRequest.Options): JSONObject {
        return JSONObject()
    }
}
